marshal.go

```diff
@@ -0,0 +1,272 @@
+// Copyright 2015 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package httprequest
+
+import (
+	"bytes"
+	"encoding"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"reflect"
+
+	"github.com/julienschmidt/httprouter"
+	"gopkg.in/errgo.v1"
+)
+
+// Marshal takes the input structure and creates an http request.
+//
+// See: Unmarshal for more details.
+//
+// For fields with a "path" item in the structural tag, the base uri must
+// contain a placeholder with its name.
+// Example:
+// For
+//    type Test struct {
+//	    username string `httprequest:"user,path"`
+//    }
+// ...the request url must contain a ":user" placeholder:
+//    http://localhost:8081/:user/files
+//
+// If a type does not implement the encoding.TextMarshaler fmt.Sprint will
+// be used to marshal its value.
+func Marshal(baseURL, method string, x interface{}) (*http.Request, error) {
+	xv := reflect.ValueOf(x)
+	pt, err := getRequestType(xv.Type())
+	if err != nil {
+		return nil, errgo.WithCausef(err, ErrBadUnmarshalType, "bad type %s", xv.Type())
+	}
+	req, err := http.NewRequest(method, baseURL, bytes.NewBuffer(nil))
+	if err != nil {
+		return nil, errgo.Mask(err)
+	}
+	p := &Params{req, httprouter.Params{}}
+	if err := marshal(p, xv, pt); err != nil {
+		return nil, errgo.Mask(err, errgo.Is(ErrUnmarshal))
+	}
+	return p.Request, nil
+}
+
+// marshal is the internal version of Marshal.
+func marshal(p *Params, xv reflect.Value, pt *requestType) error {
+	if xv.Kind() == reflect.Ptr {
+		xv = xv.Elem()
+	}
+	for _, f := range pt.fields {
+		fv := xv.FieldByIndex(f.index)
+
+		// TODO store the field name in the field so
+		// that we can produce a nice error message.
+		if err := f.marshal(fv, p); err != nil {
+			return errgo.WithCausef(err, ErrUnmarshal, "cannot marshal field")
+		}
+	}
+
+	urlString := p.URL.Path
+	var pathBuffer bytes.Buffer
+	paramsByName := make(map[string]string)
+	for _, param := range p.PathVar {
+		paramsByName[param.Key] = param.Value
+	}
+
+	offset := 0
+	hasParams := false
+	for i := 0; i < len(urlString); i++ {
+		c := urlString[i]
+		if c != ':' {
```

Comment:
if c != ':' && c != '*" {

(we should allow the * form too, as supported by httprouter).